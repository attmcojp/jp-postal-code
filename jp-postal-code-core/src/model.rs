/// 郵便番号レコード
///
/// オリジナルの情報を残すようなモデルにしている。
///
/// # Example
/// ```rust
/// # use jp_postal_code_core::model::UtfKenAllRecord;
/// let _ = UtfKenAllRecord {
///     local_government_code: "01101".to_string(),
///     old_postal_code: "060  ".to_string(),
///     postal_code: "0600000".to_string(),
///     prefecture_kana: "ホッカイドウ".to_string(),
///     city_kana: "サッポロシチュウオウク".to_string(),
///     town_kana: "イカニケイサイガナイバアイ".to_string(),
///     prefecture: "北海道".to_string(),
///     city: "札幌市中央区".to_string(),
///     town: "以下に掲載がない場合".to_string(),
///     has_multi_postal_code: 0,
///     has_chome: 0,
///     has_multi_town: 0,
///     update_code: 0,
///     update_reason: 0,
/// };
/// ```
///
/// # References
/// - [郵便番号データ（1レコード1行、UTF-8形式）の説明](https://www.post.japanpost.jp/zipcode/dl/utf-readme.html)
///
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub struct UtfKenAllRecord {
    /// 全国地方公共団体コード（JIS X0401、X0402）
    pub local_government_code: String,
    /// 旧郵便番号（5桁）
    pub old_postal_code: String,
    /// 郵便番号（7桁）
    pub postal_code: String,
    /// 都道府県名（カタカナ）
    pub prefecture_kana: String,
    /// 市区町村名（カタカナ）
    pub city_kana: String,
    /// 町域名（カタカナ）
    pub town_kana: String,
    /// 都道府県名
    pub prefecture: String,
    /// 市区町村名
    pub city: String,
    /// 町域名
    pub town: String,
    /// 一町域が二以上の郵便番号で表される場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_multi_postal_code: i16,
    /// 丁目を有する町域の場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_chome: i16,
    /// 一つの郵便番号で二以上の町域を表す場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_multi_town: i16,
    /// 更新の表示
    ///
    /// - 0: 変更なし
    /// - 1: 変更あり
    /// - 2: 廃止
    pub update_code: i16,
    /// 更新理由
    ///
    /// - 0: 変更なし
    /// - 1: 市政・区政・町政・分区・政令指定都市施行
    /// - 2: 住居表示の実施
    /// - 3: 区画整理
    /// - 4: 郵便区調整等
    /// - 5: 訂正
    /// - 6: 廃止
    pub update_reason: i16,
}
