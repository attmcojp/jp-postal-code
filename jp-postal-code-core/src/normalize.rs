mod zenkaku;

// 正規表現の定義用マクロ
macro_rules! define {
    ($name:ident, $re:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($re).unwrap());
    };
}
#[macro_use]
mod town;

pub use town::normalize_utf_ken_all_record_town;
