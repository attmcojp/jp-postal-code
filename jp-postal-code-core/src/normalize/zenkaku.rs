use std::sync::LazyLock;

/// 全角数字キャラクターか判定する
pub fn is_zenkaku_numeric(c: &char) -> bool {
    matches!(*c, '０'..='９')
}

/// 文字列から全角数字を抽出する
pub fn extract_zenkaku_numeric(s: &str) -> String {
    s.chars()
        .skip_while(|c| !is_zenkaku_numeric(c))
        .take_while(is_zenkaku_numeric)
        .collect()
}

/// 全角数字列を数値に変換する
pub fn from_zenkaku_numeric(s: &str) -> usize {
    let n: String = extract_zenkaku_numeric(s)
        .chars()
        .map(|c| match c {
            '０' => '0',
            '１' => '1',
            '２' => '2',
            '３' => '3',
            '４' => '4',
            '５' => '5',
            '６' => '6',
            '７' => '7',
            '８' => '8',
            '９' => '9',
            _ => unreachable!(),
        })
        .collect();
    n.parse()
        .unwrap_or_else(|e| panic!("Failed to parse {}: {}", s, e))
}

/// 数値を全角数字列に変換する
pub fn to_zenkaku_numeric(n: usize) -> String {
    n.to_string()
        .chars()
        .map(|c| match c {
            '0' => '０',
            '1' => '１',
            '2' => '２',
            '3' => '３',
            '4' => '４',
            '5' => '５',
            '6' => '６',
            '7' => '７',
            '8' => '８',
            '9' => '９',
            _ => unreachable!(),
        })
        .collect()
}

/// 全角で表現されたラベル付きの範囲をリスト化する
pub fn zenkaku_range_label(start: &str, end: &str) -> Vec<String> {
    static PATTERN: LazyLock<regex::Regex> =
        LazyLock::new(|| regex::Regex::new(r"[０-９]+").unwrap());
    let s = from_zenkaku_numeric(start);
    let e = from_zenkaku_numeric(end);
    (s..=e)
        .map(|n| PATTERN.replace(start, to_zenkaku_numeric(n)).to_string())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_zenkaku_numeric() {
        assert!(is_zenkaku_numeric(&'０'));
        assert!(is_zenkaku_numeric(&'１'));
        assert!(is_zenkaku_numeric(&'２'));
        assert!(is_zenkaku_numeric(&'３'));
        assert!(is_zenkaku_numeric(&'４'));
        assert!(is_zenkaku_numeric(&'５'));
        assert!(is_zenkaku_numeric(&'６'));
        assert!(is_zenkaku_numeric(&'７'));
        assert!(is_zenkaku_numeric(&'８'));
        assert!(is_zenkaku_numeric(&'９'));
        assert!(!is_zenkaku_numeric(&'あ'));
        assert!(!is_zenkaku_numeric(&'ア'));
        assert!(!is_zenkaku_numeric(&'一'));
        assert!(!is_zenkaku_numeric(&'1'));
        assert!(!is_zenkaku_numeric(&'a'));
    }

    #[test]
    fn test_extract_zenkaku_numeric() {
        assert_eq!(extract_zenkaku_numeric("１２３"), "１２３");
        assert_eq!(extract_zenkaku_numeric("１２３abc"), "１２３");
        assert_eq!(extract_zenkaku_numeric("abc１２３"), "１２３");
        assert_eq!(extract_zenkaku_numeric("abc１２３abc"), "１２３");
        assert_eq!(extract_zenkaku_numeric("abc１２３abc４５６"), "１２３");
    }

    #[test]
    fn test_from_zenkaku_numeric() {
        assert_eq!(from_zenkaku_numeric("１２３"), 123);
        assert_eq!(from_zenkaku_numeric("１２３abc"), 123);
        assert_eq!(from_zenkaku_numeric("abc１２３"), 123);
        assert_eq!(from_zenkaku_numeric("abc１２３abc"), 123);
        assert_eq!(from_zenkaku_numeric("abc１２３abc４５６"), 123);
    }

    #[test]
    fn test_to_zenkaku_numeric() {
        assert_eq!(to_zenkaku_numeric(123), "１２３");
    }

    #[test]
    fn test_range_label() {
        assert_eq!(zenkaku_range_label("１", "３"), vec!["１", "２", "３"]);
        assert_eq!(
            zenkaku_range_label("８丁目", "１２丁目"),
            vec!["８丁目", "９丁目", "１０丁目", "１１丁目", "１２丁目"]
        );
        assert_eq!(
            zenkaku_range_label("第８地割", "第１２地割"),
            vec![
                "第８地割",
                "第９地割",
                "第１０地割",
                "第１１地割",
                "第１２地割"
            ]
        );
    }
}
