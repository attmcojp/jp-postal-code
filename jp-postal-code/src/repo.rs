use jp_postal_code_core::model::UtfKenAllRecord;

pub const DEFAULT_SEARCH_PAGE_SIZE: usize = 10;

#[derive(Debug, Clone)]
pub struct UtfKenAllRepositorySearchRequest<'a> {
    pub postal_code: &'a str,
    pub page_size: Option<usize>,
    pub page_token: Option<&'a str>,
}

#[derive(Debug, Clone)]
pub struct UtfKenAllRepositorySearchResponse {
    pub records: Vec<UtfKenAllRecord>,
    pub next_page_token: Option<String>,
}

/// 郵便番号データベースを扱うリポジトリ
pub trait UtfKenAllRepository: Clone + Send + Sync {
    type Error: std::error::Error + Send + Sync + 'static;

    /// 郵便番号データベースを置き換える
    fn replace(
        &mut self,
        records: &[UtfKenAllRecord],
    ) -> impl std::future::Future<Output = Result<(), Self::Error>> + Send;

    /// 郵便番号データベースから前方一致でレコードを検索する
    fn search(
        &self,
        req: UtfKenAllRepositorySearchRequest<'_>,
    ) -> impl std::future::Future<Output = Result<UtfKenAllRepositorySearchResponse, Self::Error>> + Send;

    /// 郵便番号データベースの総数をカウントする
    fn count(&self) -> impl std::future::Future<Output = Result<usize, Self::Error>> + Send;
}
