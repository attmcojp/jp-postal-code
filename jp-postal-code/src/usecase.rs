use crate::repo::UtfKenAllRepository;
use jp_postal_code_core::model::UtfKenAllRecord;
use jp_postal_code_core::normalize::normalize_utf_ken_all_record_town;
use jp_postal_code_util::{download, parse_utf_ken_all_zip, UTF_KEN_ALL_URL};

/// 郵便番号データベースを更新する
#[tracing::instrument(skip(repo, utf_ken_all_zip_url))]
pub async fn update_postal_code_database<R, S>(
    repo: &mut R,
    utf_ken_all_zip_url: Option<S>,
) -> Result<(), anyhow::Error>
where
    R: UtfKenAllRepository,
    S: Into<String>,
{
    let utf_ken_all_zip_url = {
        if let Some(utf_ken_all_zip_url) = utf_ken_all_zip_url {
            utf_ken_all_zip_url.into()
        } else {
            UTF_KEN_ALL_URL.into()
        }
    };
    let mut tempfile = tempfile::tempfile()?;
    tracing::info!(
        ?utf_ken_all_zip_url,
        ?tempfile,
        "Download utf_ken_all.zip into a temp file"
    );
    download(utf_ken_all_zip_url, &mut tempfile).await?;
    tracing::info!(?tempfile, "Parse utf_ken_all.zip to records");
    let records = parse_utf_ken_all_zip(tempfile)?;
    tracing::info!(
        record_count = records.len(),
        "Replace database with the new records"
    );
    repo.replace(&records).await?;
    Ok(())
}

/// 郵便番号を検索する
#[tracing::instrument(skip(repo, postal_code))]
pub async fn search_postal_code<R>(
    repo: &R,
    postal_code: impl AsRef<str>,
) -> Result<Vec<UtfKenAllRecord>, anyhow::Error>
where
    R: UtfKenAllRepository,
{
    let postal_code = postal_code.as_ref();
    let records = repo.search(postal_code).await?;
    let records = records
        .into_iter()
        .flat_map(|r| {
            let towns = normalize_utf_ken_all_record_town(&r);
            towns
                .into_iter()
                .map(|town| UtfKenAllRecord { town, ..r.clone() })
                .collect::<Vec<_>>()
        })
        .collect();
    Ok(records)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::infra::ephemeral::UtfKenAllRepositoryEphemeral;

    #[tokio::test]
    async fn test_update_utf_ken_all_database() {
        let mut server = mockito::Server::new_async().await;
        let mock = server
            .mock("GET", "/zipcode/dl/utf/zip/utf_ken_all.zip")
            .with_status(200)
            .with_header("content-type", "application/zip")
            .with_body_from_file("./testdata/partial_utf_ken_all.zip")
            .create_async()
            .await;

        let mut repo = UtfKenAllRepositoryEphemeral::default();
        update_postal_code_database(
            &mut repo,
            Some(server.url() + "/zipcode/dl/utf/zip/utf_ken_all.zip"),
        )
        .await
        .unwrap();
        mock.assert_async().await;
        insta::assert_debug_snapshot!(repo.into_inner().lock().unwrap());
    }

    #[tokio::test]
    async fn test_search_postal_code() {
        let repo = UtfKenAllRepositoryEphemeral::new(vec![
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600000".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "イカニケイサイガナイバアイ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "以下に掲載がない場合".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 0,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640941".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "アサヒガオカ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "旭ケ丘".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600041".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリヒガシ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通東".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600042".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（１〜１９丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640820".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（２０〜２８丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
        ]);
        let records = search_postal_code(&repo, "064").await.unwrap();
        insta::assert_debug_snapshot!(records);
    }
}
