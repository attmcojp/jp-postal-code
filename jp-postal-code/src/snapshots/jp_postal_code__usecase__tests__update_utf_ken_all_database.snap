---
source: jp-postal-code/src/usecase.rs
expression: repo.into_inner().lock().unwrap()
---
[
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600000",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "イカニケイサイガナイバアイ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 0,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640941",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "アサヒガオカ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "旭ケ丘",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600041",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリヒガシ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通東",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西３丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西４丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西５丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西６丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西７丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西８丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西９丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１０丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１１丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１２丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１３丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１４丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１５丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１６丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１７丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１８丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西１９丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２０丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２１丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２２丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２３丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２４丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２５丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２６丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２７丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西２８丁目",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
]
